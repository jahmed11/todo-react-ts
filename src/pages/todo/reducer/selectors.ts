import { createSelector } from "@reduxjs/toolkit";
import { RootState } from "store";


const todo = (state: RootState) => state.todo;

export const todoItemsSelector  = createSelector(todo, (todoState) => todoState.todoItems);
