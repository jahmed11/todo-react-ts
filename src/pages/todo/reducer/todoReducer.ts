import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { TodoProps } from 'types/todo';


interface TodoState {
    todoItems: TodoProps[];
}

export const initialState: TodoState = {
  todoItems: [],
};

export const todoSlice = createSlice({
  name: 'todos',
  initialState,
  reducers: {
    addTodo: (state, action: PayloadAction<string>) => {
      const newTodo: TodoProps = {
        id: Date.now(),
        text: action.payload,
      };
      state.todoItems.push(newTodo);
    },
    deleteTodo: (state, action: PayloadAction<number>) => {
      state.todoItems = state.todoItems.filter(todo => todo.id !== action.payload);
    },
  },
});


export const { addTodo, deleteTodo } = todoSlice.actions;

export default todoSlice.reducer;
