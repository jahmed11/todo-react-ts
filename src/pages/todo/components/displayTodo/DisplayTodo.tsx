import { useAppSelector, useAppDispatch } from "hooks/reduxHooks";
import { deleteTodo } from "pages/todo/reducer/todoReducer";
import { sortedTodoItemsSelector } from "pages/todo/reducer/selectors";
import Todo from "./Todo";
import styles from "./displayTodo.module.css";
import { TodoProps } from "types/todo";

const DisplayTodo = () => {
  const dispatch = useAppDispatch();
  const todoItems = useAppSelector(sortedTodoItemsSelector);


  const onTodoDeleteClick = (id: number) => {
    dispatch(deleteTodo(id));
  };


  const renderTodoList = (items: TodoProps[]) => {
    if (!items.length) {
      return <div className={styles["todos-list--empty"]}>Please Add your todos</div>;
    }

    return items.map((todoItem: TodoProps, index: number) => {
      return (
        <Todo
          key={todoItem.id}
          index={index}
          text={todoItem.text}
          id={todoItem.id}
          onTodoDeleteClick={onTodoDeleteClick}
        />
      );
    });
  };

  return (
    <div className={styles["todos-container"]}>
      <h2>Todo List</h2>
      <div className={styles["todos-list"]}>{renderTodoList(todoItems)}</div>
    </div>
  );
};

export default DisplayTodo;
