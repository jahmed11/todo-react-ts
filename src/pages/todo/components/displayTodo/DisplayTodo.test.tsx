import { screen, fireEvent } from '@testing-library/react';
import { renderWithProviders } from 'utils/testUtils';
import DisplayTodo from './DisplayTodo';
import { deleteTodo } from 'pages/todo/reducer/todoReducer';

jest.mock('./Todo', () => ({
  __esModule: true,
  default: ({ text, id, onTodoDeleteClick }: any) => (
    <div>
      <span>{text}</span>
      <button onClick={() => onTodoDeleteClick(id)}>Delete</button>
    </div>
  ),
}));

describe('DisplayTodo Component', () => {
  const preloadedState = {
    todo: {
      todoItems: [
        { id: 1, text: 'Todo 1' },
        { id: 2, text: 'Todo 2' },
        { id: 3, text: 'Todo 3' },
        { id: 4, text: 'Todo 4' },
      ],
    },
  };

  it('renders correctly with todos and matches the snapshot', () => {
    const { asFragment } = renderWithProviders(<DisplayTodo />, { preloadedState });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the correct number of Todo components based on todoItems length', () => {
    renderWithProviders(<DisplayTodo />, { preloadedState });

    const todoItems = screen.getAllByText(/Todo \d/);
    expect(todoItems.length).toBe(4);
  });

  it('calls onTodoDeleteClick with the correct id when delete button is clicked', () => {
    const mockDispatch = jest.fn();
    jest.spyOn(require('hooks/reduxHooks'), 'useAppDispatch').mockReturnValue(mockDispatch);

    renderWithProviders(<DisplayTodo />, { preloadedState });

    const deleteButtons = screen.getAllByText('Delete');
    fireEvent.click(deleteButtons[0]);

    expect(mockDispatch).toHaveBeenCalledWith(deleteTodo(1));
  });
});
