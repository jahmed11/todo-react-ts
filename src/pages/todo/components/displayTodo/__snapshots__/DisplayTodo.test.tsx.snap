// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayTodo Component renders correctly with todos and matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="todos-container"
  >
    <h2>
      Todo List
    </h2>
    <div
      class="todos-list"
    >
      <div
        class="todo-item"
      >
        <span
          class="todo-item__number"
        >
          1.
        </span>
        <span
          class="todo-item__text"
        >
          Todo 1
        </span>
        <span
          class="todo-item__action"
        >
          <button
            class="button button--danger"
          >
            Delete
          </button>
        </span>
      </div>
      <div
        class="todo-item"
      >
        <span
          class="todo-item__number"
        >
          2.
        </span>
        <span
          class="todo-item__text"
        >
          Todo 2
        </span>
        <span
          class="todo-item__action"
        >
          <button
            class="button button--danger"
          >
            Delete
          </button>
        </span>
      </div>
      <div
        class="todo-item"
      >
        <span
          class="todo-item__number"
        >
          3.
        </span>
        <span
          class="todo-item__text"
        >
          Todo 3
        </span>
        <span
          class="todo-item__action"
        >
          <button
            class="button button--danger"
          >
            Delete
          </button>
        </span>
      </div>
      <div
        class="todo-item"
      >
        <span
          class="todo-item__number"
        >
          4.
        </span>
        <span
          class="todo-item__text"
        >
          Todo 4
        </span>
        <span
          class="todo-item__action"
        >
          <button
            class="button button--danger"
          >
            Delete
          </button>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
